File: Assets/Scripts/Domain/Defs/ActionDef.cs
using UnmappedIsland.Domain.Runtime;

namespace UnmappedIsland.Domain.Defs
{
    /// <summary>showMenuの値（11.1節）。現時点ではalwaysのみ（ActionSystem.md 7節）。</summary>
    public enum ShowMenuMode
    {
        Always,
    }

    /// <summary>
    /// メニュー型の宣言的操作（GameElementDefinition.md 11節）。conditionsと条件成立時の効果を1つの定義として持つ。
    /// </summary>
    public sealed class ActionDef
    {
        public string Name { get; }
        public ShowMenuMode ShowMenu { get; }

        /// <summary>nullなら常に真（conditions省略）。</summary>
        private readonly ConditionNode conditions;

        /// <summary>条件成立時に適用する効果。nullなら何も起きない。</summary>
        private readonly ActiveEffect effect;

        /// <summary>実行にかかるゲーム内時間（分）。リテラルか{object, prop}参照（weightの10.2節と同じ二択）。
        /// nullなら時間を消費しない。時間進行（AdvanceWorldTime）まではこのActionDefの責務で、
        /// 呼び出し側が実行後に別途時間を進める必要はない。</summary>
        private readonly WeightSpec? duration;

        public ActionDef(
            string name,
            ShowMenuMode showMenu,
            ConditionNode conditions,
            ActiveEffect effect,
            WeightSpec? duration = null)
        {
            Name = name;
            ShowMenu = showMenu;
            this.conditions = conditions;
            this.effect = effect;
            this.duration = duration;
        }

        public bool TryExecute(WorldObject self, WorldObject actor, WorldSession session)
        {
            if (conditions != null && !conditions.Evaluate(root => ReferenceRootResolver.Resolve(root, self, actor, dragged: null)))
                return false;

            // 時間進行はeffect適用の後（先に進めるとtick中のdestroy等がselfを破棄してから効果を適用する事故になる）。
            // ただし参照durationは適用前のselfから読む必要があるため、解決だけは適用前に行う。
            int minutes = duration.HasValue ? (int)duration.Value.Resolve(self, actor, dragged: null) : 0;

            if (effect != null) self.ApplyActiveEffect(effect, session, actor, dragged: null);

            // Worldを持たないセッション（単体テスト等、時間の概念が無い文脈）では時間進行をスキップする。
            if (minutes > 0 && session.World != null) session.AdvanceWorldTime(minutes);
            return true;
        }
    }
}

File: src/domain/defs/ActionDef.ts
import type { WorldObject } from '../runtime/WorldObject';
import type { WorldSession } from '../runtime/WorldSession';
import type { ActiveEffect } from './ActiveEffect';
import type { DefNames, DescriptionWriter } from './Description';
import { text } from './Description';
import { InteractionDef } from './InteractionDef';
import type { WeightSpec } from './PickEffect';
import type { Requirement, Requirements } from './Requirement';

/**
 * showMenuの値（11.1節）。`never`は**画面のボタンには出さない操作**で、起こすのは時間の側になる
 * ——動物の1手（docs/engine/HuntingSystem.md 5節）は名前で指して実行される点だけがアクションと
 * 同じで、プレイヤーが押す機会は無い。
 */
export type ShowMenuMode = 'always' | 'never';

/**
 * メニュー型の宣言的操作（GameElementDefinition.md 11節）。1枚のカード（self）だけで完結し、
 * 名前で指して実行される。actorは常に暗黙的に参加する。
 */
export class ActionDef extends InteractionDef {
  readonly showMenu: ShowMenuMode;

  constructor(
    name: string,
    showMenu: ShowMenuMode,
    requirements: Requirements | undefined,
    effect: ActiveEffect,
    duration?: WeightSpec,
  ) {
    super(name, requirements, effect, duration);
    this.showMenu = showMenu;
  }

  protected describeTrigger(_names: DefNames, out: DescriptionWriter): void {
    out.write(text(`show_menu: ${this.showMenu}`));
  }

  get kind(): 'action' {
    return 'action';
  }

  tryExecute(self: WorldObject, actor: WorldObject | undefined, session: WorldSession): boolean {
    return this.apply(self, undefined, actor, session);
  }

  /** 今このアクションを実行できない理由（最初に落ちた要件）。実行できるならundefined。 */
  unmetRequirement(self: WorldObject, actor: WorldObject | undefined): Requirement | undefined {
    return this.firstUnmetRequirement(self, undefined, actor);
  }
}
